test/http_memprof/http_memprof.go

```diff
@@ -0,0 +1,278 @@
+	allocDiffMb := bytesToMegabytes(allocDiff)
+
+	fmt.Printf("%v used %.2f MB and returned a response of length %d\n", key, allocDiffMb, len(respStr))
+	return &MemoryProfileResult{
+		profileKey:     key,
+		allocMB:        allocDiffMb,
+		responseLength: len(respStr),
+	}
+
+	// NOTE: Another approach here would be to use go tool pprof to compute a
+	// profile with "substraction", where the profile consists of the
+	// differences with the "base profile". The numerical diff can be read from
+	// that "diff profile" the same way we read the total memory allocated in
+	// the "before" and "after" profiles currently.
+	// To achieve this, one could do something like the following:
+	//////////
+	// profDiff := fmt.Sprintf(profileLocationTemplate, "diff", key)
+	// outputFlag := fmt.Sprintf("-output=%v", profDiff)
+	// // error-checking omitted from the following
+	// cmd := exec.Command("go", "tool", "pprof", "-proto", outputFlag, "-sample_index=alloc_space", "-base", profBefore, profAfter)
+	// allocDiffMb := bytesToMegabytes(GetTotalSpaceAllocFromProfile(profDiff)
+	//////////
+}
+
+func writeResultsToCsv(results []*MemoryProfileResult, outputPath string) {
+	f, err := os.Create(outputPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+
+	header := "ProfileKey,AllocatedMemoryMB,ResponseLength"
+	fmt.Fprintln(f, header)
+
+	for _, result := range results {
+		fmt.Fprintf(f, "%v,%.2f,%d\n", (*result).profileKey, (*result).allocMB, (*result).responseLength)
+	}
+}
+
+func main() {
+	log.SetFlags(log.LstdFlags | log.Lshortfile)
+
+	flag.Parse()
+
+	// Create the output folder, if it doesn't exist.
+	// Form the actual output folder as <flag_value>/<unix_time> to preserve
+	// data from previous runs
+	now := time.Now()
+	outFolder := filepath.Join(*outFolderParent, strconv.FormatInt(now.Unix(), 10))
+	fmt.Println("Writing results to path:", outFolder)
+	err := os.MkdirAll(outFolder, 0o0755)
+	if err != nil {
+		log.Fatalf("could not create temp directory at %v: %v", outFolder, err)
+	}
+
+	// Set up a connection to the server.
+	// gRPC server start listening last, so if we have connected to gRPC, we can
+	// assume that the pprof HTTP handlers is also on.
+	fmt.Printf("Attempting to make a connection to the gRPC server at %v\n", *grpcAddr)
+	conn, err := grpc.Dial(*grpcAddr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(100000000 /* 100M */)),
+		// The grpc.WithBlock() option make this call block until a connection is
+		// established, internally retrying until needed.
+		grpc.WithBlock(),
+	)
+	if err != nil {
+		log.Printf("Could not connect: %v\n", err)
+	}
+	log.Println("Connected to gRPC succesfully")
+
+	defer conn.Close()
+	c := pb.NewMixerClient(conn)
+	ctx := context.Background()
+
+	var profileResults []*MemoryProfileResult
+	var count int64
+	for _, allFacets := range []bool{true, false} {
+		for _, r := range []struct {
```

Comment:
Can you add a TODO here to move these into a config based file to incorporate for more APIs.